src/Makefile

```diff
@@ -267,6 +267,17 @@ ifeq ($(MALLOC),jemalloc)
 	FINAL_LIBS := ../deps/jemalloc/lib/libjemalloc.a $(FINAL_LIBS)
 endif
 
+# Detect if the OS supports posix_fadvise.
+HAVE_FADVISE := $(shell sh -c ' \
```

Comment:
See the discussion https://github.com/redis/redis/pull/11248#discussion_r1051589528, at first I also detect it by platform, but @oranagra remind me that the support may not only depend on platform but also the kernel version, so I deside to seek for compile-time detection. I think this also helps on some novel syscall, e.g. iouring and so on.

Reply:
@yossigo so can we keep the current code?
i do wanna support any OS that has it, not just linux.
or maybe you argue that we should empirically find the OSes we support that have it and use it only on these (leaving others out of scope)?

Reply:
@oranagra Yes, I think we can white list the specific platforms on which we assume it exists and ignore it elsewhere. It's more aligned with how most other capabilities are handled.

Reply:
i think other things that we handle that way are non-standard (like `sync_file_range`), but `posix_fadvise` is standard (like `_Atomic`.
so using it on just a few platforms (e.g. just Linux and Mac), could theoretically be a wasteful miss on Solaris (just random examples, i didn't bother to check which one supports it).

on the other hand, we can say that since this is just an advise, and we're not sure what's the implementation behind each os, we don't wanna use it on ones on which we didn't test it..
although, we are never gonna test it on these, so if we'll enable it, we can get some report of issues from the community, but if we're wasteful on some of these oses because we didn't bother to enable it, no one will report that.

did i convince you? or do you insist on explicitly enabling it on specific list of OSes?

Reply:
Personally I perfer autoconf-like detection since it's more robust and portable, and doesn't depend on the programer's good understanding to the environment as well as its history. I refer to some `make` project, like nginx, rocksdb, they detect the symbols with shell script, while other `cmake` project utilize the cmake mechanisms like `check_symbol_exists` and `check_cxx_source_compiles`. In redis's tradition these detection were done with macros, so it involves some break...

Reply:
I discussed this with Yossi and decided to drop it in favor of ifdefs in config.h.
personally i'm not very happy about it, but willing to let go.
the arguments were that we don't do such a thing for many other system calls, and we don't want to transition into adding more of these in the current way. arguing that the _Atomic thing was an exception because the problem was rather widespread.

putting this in config.h we have two alternatives:
1. by default try to use it, and just add an explicit set of exception OSes on which it's missing.
2. by default don't use it, and add exceptions only for the OSes on which we know it exists and works as expected.

the disadvantage of [1] would be that compilation would break on some system, and then we'll get a PR to add an exception.
the disadvantage of [2] is that on some system were we could have performed better, we won't and it's likely no one will notice.
we decided to take option [2].

@uvletter i know you've invested quite a lot of time on this mechanism, i apologize for pushing in that direction and wasting your time.

Reply:
@oranagra  It's just a process of exploration, so no need for the apologize, and I think I learned a lot from the discussion : )

I'll update this part tomorrow. I also spent sometime updating the top comment, but since I'm a native speaker, I'm sure if I describe correctly. If there're some faults or other points I missed, plz help me refine it thanks!